tests/bot/cogs/moderation/test_utils.py

```diff
@@ -0,0 +1,477 @@
+                    "hidden": False,
+                    "reason": "Test Ban",
+                    "type": "note",
+                    "user": self.member.id,
+                    "active": True
+                }
+            },
+            {
+                "args": (self.ctx, self.member, "mute", "Test Ban"),
+                "expected_output": None,
+                "raised_error": ResponseCodeError(AsyncMock(status=400), {'user': 1234}),
+                "payload": {
+                    "actor": self.ctx.message.author.id,
+                    "hidden": False,
+                    "reason": "Test Ban",
+                    "type": "mute",
+                    "user": self.member.id,
+                    "active": True
+                }
+            },
+            {
+                "args": (self.ctx, self.member, "ban", "Test Ban", now, True, False),
+                "expected_output": [
+                    {
+                        "id": 1,
+                        "inserted_at": "2018-11-22T07:24:06.132307Z",
+                        "expires_at": "5018-11-20T15:52:00Z",
+                        "active": True,
+                        "user": 1234,
+                        "actor": 1234,
+                        "type": "ban",
+                        "reason": "Test Ban",
+                        "hidden": False
+                    }
+                ],
+                "raised_error": None,
+                "payload": {
+                    "actor": self.ctx.message.author.id,
+                    "hidden": True,
+                    "reason": "Test Ban",
+                    "type": "ban",
+                    "user": self.member.id,
+                    "active": False,
+                    "expires_at": now.isoformat()
+                }
+            },
+        ]
+
+        for case in test_cases:
+            args = case["args"]
+            expected = case["expected_output"]
+            raised = case["raised_error"]
+            payload = case["payload"]
+
+            with self.subTest(args=args, expected=expected, raised=raised, payload=payload):
+                self.ctx.bot.api_client.post.reset_mock(side_effect=True)
+                post_user_mock.reset_mock()
+
+                if raised:
+                    self.ctx.bot.api_client.post.side_effect = raised
+
+                post_user_mock.return_value = "foo"
+
+                self.ctx.bot.api_client.post.return_value = expected
+
+                result = await post_infraction(*args)
+
+                self.assertEqual(result, expected)
+
+                if not raised:
+                    self.bot.api_client.post.assert_awaited_once_with("bot/infractions", json=payload)
+
+                if hasattr(raised, "status") and hasattr(raised, "response_json"):
+                    if raised.status == 400 and "user" in raised.response_json:
+                        post_user_mock.assert_awaited_once_with(args[0], args[1])
```

Comment:
This doesn't test the control flow for the loop it uses with `should_post_user`. Here's an outline of the behaviour you need to test:

- [x] Post the infraction
- [x] If the infraction post fails cause the user isn't in the db, post the user
- [ ] If the user post fails, exit out of the loop right away
- [ ] If the user post succeeds, attempt to post the infraction again
- [ ] If the infraction fails to post _again_, then do nothing and exit the loop
- [ ] If the error is something besides a 400, send an error message and exit the loop

Reply:
How I should do all these other checks? It's confusing for me.

Reply:
For the third point, consider that when the loop repeats, `post()` gets called. It is undesirable for that to be repeated, therefore I phrased it as the loop being exited. But really, think about it as "assert that `post()` doesn't happen again". Such an assertion can be made by looking at the `call_count` attribute of the mock.

When a `post()` is successful, the test should ensure that no error message is sent and no user is posted. If the aforementioned assertions are made, then it effectively means the loop exited at that point. This should also be done when the `post()` fails the _second_ time too (i.e. the case in fifth point above). In most of the outlined cases, you'll want to have assertions for `call_count` to make sure it's not called too few or too many times.